include/fmt/format.h

```diff
@@ -643,7 +646,7 @@ FMT_CONSTEXPR void for_each_codepoint(string_view s, F f) {
     auto error = 0;
     auto end = utf8_decode(buf_ptr, &cp, &error);
     bool result = f(error ? invalid_code_point : cp,
-                    string_view(ptr, to_unsigned(end - buf_ptr)));
+                    string_view(ptr, error ? 1 : to_unsigned(end - buf_ptr)));
     return result ? end : nullptr;
```

Comment:
could you help me to understand in the example of "\xF0\x28", is this end still buf_ptr+4 even tho there is an error?

Reply:
@stevenzzzz 
No. On invalid utf-8 sequence, processing is one byte at a time.

Please describe your testing process.

My environment:
macOS 12.5.1
Apple clang version 13.1.6 (clang-1316.0.21.2.5)

Test steps:
```
CXXFLAGS="-fsanitize=address" cmake -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_STANDARD=20 -DFMT_PEDANTIC=ON -DFMT_WERROR=ON /path/to/patched/fmt
make
./bin/ranges-test
```
The test shows no errors.

String "\xF0\x28" present in updated test.


Without this PR:
```
ERROR: AddressSanitizer: negative-size-param: (size=-2)
```

Reply:
I might have overlooked something, but for "\xF0\x28\0\0anything" the utf8_decode_new/decode would return 4 (as expected), when the error is set to 42.  Is that "4" gonna skip the "\x28\0\0"?

same test:  https://godbolt.org/z/EbeEex4Gf

Reply:
thanks for working on this!

We got the ASAN error by printing a protobuf debugstring() which is arbitrary binary bytes, unfortunately I can't show the protobuf to you. 


I am a little bit lost on how the printing works, could you help me to understand where is the "skip one byte on error" pls?  per this piece of code: 
```
    for (auto end = p + s.size() - block_size + 1; p < end;) {
      p = decode(p, p);
      if (!p) return;
    }
```
it seems p would advance 4 bytes when encode(p, p) is called on "\xof\x28".

Reply:
@stevenzzzz 
Please replace
```c++
  auto decode = [f](const char* buf_ptr, const char* ptr) {
    auto cp = uint32_t();
    auto error = 0;
    auto end = utf8_decode(buf_ptr, &cp, &error);
    bool result = f(error ? invalid_code_point : cp,
                    string_view(ptr, error ? 1 : to_unsigned(end - buf_ptr)));
    return result ? end : nullptr;
  };
```

with:
```c++
  auto decode = [f](const char* buf_ptr, const char* ptr) {
    auto cp = uint32_t();
    auto error = 0;
    auto end = utf8_decode(buf_ptr, &cp, &error);
    bool result = f(error ? invalid_code_point : cp,
                    string_view(ptr, error ? 1 : to_unsigned(end - buf_ptr)));
    return result ? (error ? buf_ptr + 1 : end) : nullptr;
  };
```

and run test.

Reply:
I was wondering why you didn't replace that in this PR.

Reply:
@stevenzzzz 
Did this replacement fix the bug? If yes, I will update the PR.

I didn't see this bug before.